src/tshare/participant.rs

```diff
@@ -0,0 +1,1264 @@
+    #[cfg_attr(feature = "flame_it", flame("tshare"))]
+    #[instrument(skip_all, err(Debug))]
+    fn handle_ready_msg<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        message: &Message,
+    ) -> Result<ProcessOutcome<<Self as ProtocolParticipant>::Output>> {
+        info!("Handling ready tshare message.");
+
+        let ready_outcome = self.process_ready_message(rng, message)?;
+        let round_one_messages = run_only_once!(self.gen_round_one_msgs(rng, message.id()))?;
+        // extend the output with r1 messages (if they hadn't already been generated)
+        Ok(ready_outcome.with_messages(round_one_messages))
+    }
+
+    /// Generate the protocol's round one message.
+    ///
+    /// The outcome is a broadcast message containing a commitment to:
+    /// - shares [`CoeffPublic`] X_ij for all other participants,
+    /// - a "pre-commitment" A to a Schnorr proof.
+    #[cfg_attr(feature = "flame_it", flame("tshare"))]
+    #[instrument(skip_all, err(Debug))]
+    fn gen_round_one_msgs<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        sid: Identifier,
+    ) -> Result<Vec<Message>> {
+        info!("Generating round one tshare messages.");
+
+        // Generate shares for all participants.
+        let (coeff_privates, coeff_publics) = {
+            let mut privates = vec![];
+            let mut publics = vec![];
+
+            for _ in 0..self.input.threshold() {
+                let (private, public) = CoeffPublic::new_pair(rng)?;
+                privates.push(private);
+                publics.push(public);
+            }
+
+            if let Some(private) = self.input.share() {
+                privates[0] = private.clone();
+                publics[0] = private.to_public();
+            }
+
+            (privates, publics)
+        };
+
+        // Generate proof precommitments.
+        let sch_precoms = (0..coeff_publics.len())
+            .map(|_| PiSchProof::precommit(rng))
+            .collect::<Result<Vec<_>>>()?;
+
+        let decom = TshareDecommit::new(
+            rng,
+            &sid,
+            &self.id(),
+            coeff_publics,
+            sch_precoms
+                .iter()
+                .map(|sch_precom| *sch_precom.precommitment())
+                .collect(),
+        );
+
+        // Store the beginning of our proofs so we can continue the proofs later.
+        self.local_storage
+            .store::<storage::VecSchnorrPrecom>(self.id(), sch_precoms);
+
+        // Mark our own public shares as verified.
+        self.local_storage
+            .store::<storage::ValidPublicCoeffs>(self.id(), decom.coeff_publics.clone());
+
+        // Store the private share from ourselves to ourselves.
+        let my_private_share = Self::eval_private_share(&coeff_privates, self.id());
+        let my_contant_term = Self::eval_private_share_at_zero(&coeff_privates);
+        if let Some(private) = self.input.share() {
```

Comment:
We defined the field `share` as an Option, because if it is None, then we generate a random value first of all. This is not absolutely necessary, but I don't think it is a problem either. The purpose of this if here is merely to get the share from this Option, which is never None.